### ISSUE TITLE
Fails with Segmentation Fault

### ISSUE THREAD

**Issue by koorchik** [bwilcox-1234/ChatScript#37]
I've started with tutorial and then change the files for my needs.

So, I've created 
RAWDATA/filesmine.txt with the following content:
```
RAWDATA/HARRY/simplecontrol.top # default bot control
RAWDATA/TEST/tutorial.top # tutorial bot data
```

and 

RAWDATA/TEST/tutorial.top with the following content
```
concept: ~car_brand [Ford BMW Audi Chevrolet Nissan Renault Handa Tesla Toyota]
concept: ~car_model [F150 ModelS ModelX]

topic: ~introductions repeat keep []


u: ([my existing] claims) Intent: View existing claims

u: (quote for * _[~car_brand]) Intent: Get an Auto Quote. But some data is missing. $car_brand=_0
^respond(~get_car_info)

u: (quote for * _~number _~car_model) Intent: Get an Auto Quote. But some data is missing. $car_year=_0 $car_model=_1
^respond(~get_car_info)

u: (car is dead) Intent: Roadside assistance

topic: ~get_car_info repeat keep []

u: (_~car_brand) $car_brand=_0
^respond(~get_car_info)

u: (_~car_model) $car_model=_0
^respond(~get_car_info)

u: (_[~number]) $car_year=_0
^respond(~get_car_info)

u: (!$car_brand) What the brand of the car?
u: (!$car_model) What the model of the car?
u: (!$car_year) What the year of the car?

u: ($car_year $car_model $car_brand) Intent: Get an Auto Quote for $car_brand $car_model $car_year
keep() end(TOPIC)
```

then I start the ChatBot with

1. ./BINARIES/MacChatScript local
2. :build mine
3. Enter "quote for 2012 F150"

This phrase terminates the app

**Comment by solyaris** [comment]
hi @koorchik 
maybe:
* create a MINE folder

    RAWDATA/MINE/simplecontrol.top
    RAWDATA/MINE/tutorial.top

* build with:

    :build mine

BTW, good application case :-)

giorgio

**Comment by bwilcox-1234** [comment]
Classic infinite recursion blowing up the stack in C. 
topic: ~get_car_info repeat keep []
u: (_~car_brand) $car_brand=_0
^respond(~get_car_info)
You match something and then call the topic AGAIN and it matches the same thing again....

**Comment by koorchik** [comment]
@solyaris @bwilcox-1234 Thank you for quick response.

@bwilcox-1234   But why it works when I ask for "quote for my Ford"?  It goes through another template ```
u: (quote for * _[~car_brand]) Intent: Get an Auto Quote. But some data is missing. $car_brand=_0``` and it works.


Could you explain why it does match the same thing again as  ``` u: (_~car_brand) $car_brand=_0 ``` expects ~car_brand as a user input?

**Comment by solyaris** [comment]
Hi, 
I apologize if I intrude (with my ignorance)

I didn't see the recursion. 
BTW, @bwilcox-1234 maybe this could be  pointed out in build phase (possible change request).  

So the solution here could be to just remove  the ^respond(~get_car_info) in the same topic.

>Could you explain why it does match the same thing again as u: (_~car_brand) $car_brand=_0 expects >~car_brand as a user input?

the reason could be "rule precedence":  rules in introductions topic are evaluated first.

    u: (quote for * _[~car_brand]) Intent: Get an Auto Quote. But some data is missing. $car_brand=_0
^respond(~get_car_info)

here [ ] seems useless.

    u: (quote for * _~car_brand) Intent: Get an Auto Quote. But some data is missing. $car_brand=_0 ^respond(~get_car_info)

**Comment by bwilcox-1234** [comment]
ChatScript cannot know if recursion is intentional or accidental.

get_car_info has a series of rules to pick off data from input. As long as those rules do not generate output, the system AUTOMATICALLY moves on to the next rule, hence no reason to call get_car_info from inside that topic. Only the rules at the end generate output.

**Comment by solyaris** [comment]
> ChatScript cannot know if recursion is intentional or accidental.

but in CS, there is any concrete case where recursion make sense? Any example?
thanks

**Comment by bwilcox-1234** [comment]
Absolutely. But not infinite recursion. Hence CS cannot stop you or warn you that you are using recursion. Merely that if you make it infinite you blow out the stack and die. As for an example of a valid topic recursion, I can't give you one offhand. But you might have changed the current sentence internally and then wanted to use the topic again,

**Comment by solyaris** [comment]
> But you might have changed the current sentence internally and then wanted to use the topic again,

I see... maybe it could help a WARNING message, with a recursion detection during bot building phase, 
because otherwise, you know, possible segmentation faults are hard to debug in run time :)

**Comment by bwilcox-1234** [comment]
I can do that. Will be in next release.

**Comment by koorchik** [comment]
Hello everyone. Just want to share the details about the application that we were working at. My developer decided to write a post about it - http://blog.webbylab.com/how-we-built-ai-for-a-chatbot-in-javascript-and-chatscript/ I hope you will like it)
I will appreciate any feedback. And thank you for the help one more :)

**Comment by bwilcox-1234** [comment]
A fine blog. There are a number of places where you are not using CS to best advantage. 1. You dont have to alter livedata, you can in your script file use replace: xxx yyy  to get the same effect without touching a std cs release.  Also, if you do edit LIVEDATA, you dont have to rebuild your bot, merely restart the engine.  You said: :why can show full stack trace of rules and topics fired during evaluating the latest input. But that is true of :trace, not :why .    You also said: CS engine finds the first value in ~concept_A of concepts then it won’t search for it in ~concept_B.    I don't believe that. This rule says find concept and A if you can, then find concept and B if you can.  Using [] would make it pick just one of them.   Also you dont need to put your different bots on different ports. They can reside all together in a single build.   and you said when adding a misspell in livedata that you had to rebuild 0.  NOTHING in livedata requires a rebuild, merely an engine restart.

--

**Comment by solyaris** [comment]
Hi Viktor

first of all thanks to having mentioned my old article. 
And thanks for your article; it's always tiring to share a complete tutorial.

I found few minor typo here and here, by example

    u: BURGER (I [want need take] [_burger potato ice-cream])

underscore  have to  put the `_` before the `[`

    u: BURGER (I [want need take] _[burger potato ice-cream])
 
isn't it? But maybe I'm wrong.


Also:
simplecontrol.top have to be edited to get the prompt you show in examples (and maybe the bot name HARRY must be changed... that's optional ) 


Maybe you could just update a bit your article. But again I thank you for sharing!

Performance Benchmark: may you detail a bit more in deep how you did the test?

I like your idea to create subdirectories. 

Gist/request to Bruce: 
I'd like something for CS as in Ruby realms is "$ rails new"

I mean a simple command line wizard to create a new bot scaffolding.
by example

    $ ChatScript new Food

will create a `RAWDATA/Food` directory with inside a simplecontrol.top script etc. etc. 

regards
giorgio

**Comment by unsigned6** [comment]
Dear @bwilcox-1234 and @solyaris many thanks for such valuable feedback to our article. I've updated with moments you've mentioned about. I'm very excited, that it was interesting to you.

More in detail about performance test: using JS class-wrapper I was sending messages to ChatScript. In another script I've generated queue of 10K messages and started to send them asynchronously with portions (chunks) of 100 messages and measuring and saving time for each message. Then I sorted this results to get metrics showing how many percents of total messages count performance script was waiting for response from CS. If it is needed I can share with you code which I used for this test.
